### PR TITLE
Fix for GRU block that was failing fusion guard on broadcasted tensors.

### LINKED ISSUE #873: GRU Eval Issue with Fusion Guard with broadcasted tensors?
## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->
From looking at print statements in the Cuda Fusion Guard code it looks like the chunking of the Bias is somehow failing the CudaFusion Guard.  It's the 6th (index 5) input to the Guard.  I am not sure if the guard logic is wrong or if the chunking did something wrong, yet.

```
Index: 1 Size: 512 Stride: 1
Index: 0 Size: 64 Stride: 512
Index: 1 Size: 512 Stride: 1
Index: 0 Size: 64 Stride: 1536
Index: 1 Size: 512 Stride: 1
Index: 0 Size: 64 Stride: 1536
Index: 1 Size: 512 Stride: 1
Index: 0 Size: 64 Stride: 1536
Index: 1 Size: 512 Stride: 1
Index: 0 Size: 64 Stride: 1536
Index: 0 Size: 64 Stride: 0
Index: 1 Size: 512 Stride: 1
Fail2
ComplyWith False: 5 of 10
```

```
graph(%inputs.1 : Tensor,                                                                                                                                                                                                                                                                                                                                        
      %weights.1 : Tensor,                                                                                                                                                                                                                                                                                                                                       
      %bias.1 : Tensor):                                                                                                                                                                                                                                                                                                                                         
  %3 : int = prim::Constant[value=2]() # test.py:42:24                                                                                                                                                                                                                                                                                                           
  %4 : int = prim::Constant[value=3]() # test.py:41:46                                                                                                                                                                                                                                                                                                           
  %5 : int = prim::Constant[value=1]() # test.py:39:25                                                                                                                                                                                                                                                                                                           
  %6 : int = prim::Constant[value=0]() # test.py:38:29                                                                                                                                                                                                                                                                                                           
  %7 : str = prim::Constant[value="AssertionError: "]()                                                                                                                                                                                                                                                                                                          
  %8 : NoneType = prim::Constant()                                                                                                                                                                                                                                                                                                                               
  %9 : bool = prim::Constant[value=1]() # test.py:47:4                                                                                                                                                                                                                                                                                                           
  %10 : bool = prim::Constant[value=0]()                                                                                                                                                                                                                                                                                                                         
  %11 : int[] = aten::size(%inputs.1) # <string>:7:9                                                                                                                                                                                                                                                                                                             
  %timesteps.1 : int = aten::__getitem__(%11, %6) # test.py:38:16                                                                                                                                                                                                                                                                                                
  %batch.1 : int = aten::__getitem__(%11, %5) # test.py:39:12                                                                                                                                                                                                                                                                                                    
  %14 : int[] = aten::size(%weights.1) # <string>:7:9                                                                                                                                                                                                                                                                                                            
  %gru_channels.1 : int = aten::__getitem__(%14, %6) # test.py:40:19                                                                                                                                                                                                                                                                                             
  %16 : int = aten::__getitem__(%14, %5) # test.py:41:11                                                                                                                                                                                                                                                                                                         
  %17 : int = aten::mul(%gru_channels.1, %4) # test.py:41:31                                                                                                                                                                                                                                                                                                     
  %18 : bool = aten::eq(%16, %17) # test.py:41:11                                                                                                                         
   = prim::If(%18) # test.py:41:4                                                                                   
    block0():                                                                                                       
      -> ()                                                                                                       
    block1():                                                                                                      
       = prim::RaiseException(%7) # test.py:41:4                                                                   
      -> ()                                                                                                         
  %19 : int = aten::__getitem__(%11, %3) # test.py:42:11                                                                 
  %20 : bool = aten::eq(%19, %17) # test.py:42:11                                                                   
   = prim::If(%20) # test.py:42:4                                                                                        
    block0():                                                                                                     
      -> ()                                                                                                       
    block1():                                                                                                                                                                                                                                                                                                                                                    
       = prim::RaiseException(%7) # test.py:42:4                                                                                                                                                                                                                                                                                                                 
      -> ()                                                                                                                                                                                                                                                                                                                                                      
  %21 : int[] = prim::ListConstruct(%batch.1, %gru_channels.1)                                                                                                                                                                                                                                                                                                   
  %state.1 : Tensor = aten::new_zeros(%inputs.1, %21, %8, %8, %8, %8) # test.py:44:12                        
  %23 : int[] = prim::ListConstruct(%timesteps.1, %batch.1, %gru_channels.1)                                                 
  %outputs.1 : Tensor = aten::new_zeros(%inputs.1, %23, %8, %8, %8, %8) # test.py:45:14                            
  %25 : int = prim::Constant[value=8]()                                                                                       
  %26 : int = aten::__round_to_zero_floordiv(%timesteps.1, %25)                                                    
  %27 : int = aten::mul(%26, %25)                                                                                                                                        
  %28 : int = aten::sub(%timesteps.1, %27)                                                                 
  %29 : int, %state.2 : Tensor = prim::Loop(%26, %9, %6, %state.1) # test.py:47:4                                                                                         
    block0(%idx.2 : int, %32 : int, %state.14 : Tensor):                                                          
      %timestep_input.4 : Tensor = aten::select(%inputs.1, %6, %32) # test.py:48:25                               
      %35 : Tensor = aten::mm(%state.14, %weights.1) # test.py:49:26                                              
      %36 : Tensor, %37 : Tensor, %38 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%timestep_input.4)           
      %39 : Tensor[] = prim::ListConstruct(%35, %bias.1)                                                           
      %40 : Tensor[] = aten::broadcast_tensors(%39)                                                                
      %41 : Tensor, %42 : Tensor = prim::ListUnpack(%40)                                                           
      %43 : Tensor, %44 : Tensor, %45 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%41)                        
      %46 : Tensor, %47 : Tensor, %48 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%42)                         
      %49 : bool = prim::CudaFusionGuard[types=[Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], req$
ires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, devi$
e=cuda:0)]](%state.14, %38, %37, %36, %45, %48, %44, %47, %43, %46)                                                                                                                                                                                                                                                                                              
      %50 : Tensor = prim::If(%49)                                                                                                                                                                                                                                                                                                                               
        block0():                                                                                                        
          %state.56 : Tensor = prim::CudaFusionGroup_0[cache_id=0](%state.14, %38, %37, %36, %45, %48, %44, %47, %43, %46)
          -> (%state.56)                                                                                          
        block1():                                                                                                                                                                                                                                                                                                                                                
          %52 : Function = prim::Constant[name="fallback_function", fallback=1]()                                                                                                                                                                                                                                                                                
          %53 : (Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0)) = prim::CallFunction(%52, %state.14, %38, %37, %36, %45, %48, %44, %47, %43, %46)
          %54 : Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0) = prim::TupleUnpack(%53)                
          -> (%54)                                                                                                                                                        
      %55 : Tensor = aten::select(%outputs.1, %6, %32) # test.py:58:8                                                              
      %56 : Tensor = aten::copy_(%55, %50, %10) # test.py:58:8                                                     
      %57 : int = aten::add(%32, %5)                                                                             
      %timestep_input.6 : Tensor = aten::select(%inputs.1, %6, %57) # test.py:48:25                                
      %59 : Tensor = aten::mm(%50, %weights.1) # test.py:49:26                                                                                                                 
      %60 : Tensor, %61 : Tensor, %62 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%timestep_input.6)   
      %63 : Tensor[] = prim::ListConstruct(%59, %bias.1)                                                   
      %64 : Tensor[] = aten::broadcast_tensors(%63)                                                         
      %65 : Tensor, %66 : Tensor = prim::ListUnpack(%64)                     
      %67 : Tensor, %68 : Tensor, %69 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%65)
      %70 : Tensor, %71 : Tensor, %72 : Tensor = prim::ConstantChunk[chunks=3, dim=1](%66)
      %73 : bool = prim::CudaFusionGuard[types=[Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], req$
ires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, device=cuda:0), Float(64, 512, strides=[1536, 1], requires_grad=0, devi$
e=cuda:0)]](%50, %62, %61, %60, %69, %72, %68, %71, %67, %70)                                                                                                                                                                                                                                                                                                    
      %74 : Tensor = prim::If(%73)                                                                                                                                                                                                                                                                                                                               
        block0():                                                                                                 
          %state.60 : Tensor = prim::CudaFusionGroup_1[cache_id=0](%50, %62, %61, %60, %69, %72, %68, %71, %67, %70)
          -> (%state.60)
        block1():
          %76 : Function = prim::Constant[name="fallback_function", fallback=1]()
          %77 : (Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0)) = prim::CallFunction(%76, %50, %62, %61, %60, %69, %72, %68, %71, %67, %70)
          %78 : Float(64, 512, strides=[512, 1], requires_grad=0, device=cuda:0) = prim::TupleUnpack(%77)
          -> (%78)
```

## To Reproduce

```
import statistics
import time
from typing import Tuple

import torch

def run_benchmark(
    func, *inputs, n_warmup_iters=5, n_bench_iters=10
) -> Tuple[float, float]:
    with torch.no_grad():
        # Warmup
        for i in range(n_warmup_iters):
            print("Starting warmup iteration", i + 1)
            func(*inputs)

        times = []
        print(func.graph_for(*inputs))
        for i in range(n_bench_iters):
            print("Starting benchmark iteration", i + 1)
            start = time.time()
            func(*inputs)
            elapsed = time.time() - start
            print(elapsed)
            times.append(elapsed)

    return statistics.mean(times), statistics.stdev(times)


@torch.jit.script
def custom_gru_fwd(
    inputs: torch.Tensor, weights: torch.Tensor, bias: torch.Tensor
) -> torch.Tensor:
    """
    inputs: [timesteps, batch, 3 * gru_channels]
    weights: [gru_channels, 3 * gru_channels]
    bias: [3 * gru_channels]
    """
    timesteps = inputs.shape[0]
    batch = inputs.shape[1]
    gru_channels = weights.shape[0]
    assert weights.shape[1] == gru_channels * 3
    assert inputs.shape[2] == gru_channels * 3

    state = inputs.new_zeros((batch, gru_channels))
    outputs = inputs.new_zeros((timesteps, batch, gru_channels))

    for idx in range(timesteps):
        timestep_input = inputs[idx]
        recurrent_input = torch.mm(state, weights) + bias
        reset_input, update_input, new_input = torch.chunk(timestep_input, 3, dim=1)
        reset_recurrent, update_recurrent, new_recurrent = torch.chunk(
            recurrent_input, 3, dim=1
        )
        reset = torch.sigmoid(reset_input + reset_recurrent)
        update = torch.sigmoid(update_input + update_recurrent)
        new = torch.tanh(new_input + reset * new_recurrent)
        state = (1.0 - update) * new + update * state
        outputs[idx] = state
    return outputs

if __name__ == "__main__":
    print("Initializing benchmark inputs.")
    timesteps = 339
    batch = 64
    channels = 512
    inputs = torch.normal(0.0, 0.1, [timesteps, batch, 3 * channels])
    weights = torch.normal(0.0, 0.1, [channels, 3 * channels])
    bias = torch.normal(0.0, 0.1, [3 * channels])

    print("Starting GPU benchmark.")
    with torch.jit.fuser("fuser2"):
        gpu_mean, gpu_std = run_benchmark(custom_gru_fwd, inputs.cuda(), weights.cuda(), bias.cuda())
    print(f"GPU mean: {gpu_mean * 1000:.3f}")
    print(f"GPU stddev: {gpu_std * 1000:.3f}")
```

### PR DESCRIPTION
**Pull request by kevinstephano**
Fixes  #873

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/interface.cpp [hunk 1]
```diff
@@ -125,7 +125,7 @@ bool complyWith(
         // we are not looking at dim-j, but dim-sorted_index, which
         // is the j-th fastest dim;
         // TODO: merge this with above and put a long comment there
-        if (t_strides[sorted_index] < t_strides[inner_dim]) {
+        if ((t_strides[sorted_index] != 0) and (t_strides[sorted_index] < t_strides[inner_dim])) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/interface.cpp:128]: `and`?

File: torch/csrc/jit/codegen/cuda/interface.cpp [hunk 2]
```diff
@@ -142,8 +142,8 @@ bool complyWith(
                 stride_properties[j - 1]->stride_index_.has_value(),
                 "Counknown index is meaningless");
             // TODO: merge this check up
-            if (t_strides[sorted_index] !=
-                t_strides[inner_dim] * t_sizes[inner_dim]) {
+            if ((t_strides[sorted_index] != 0) and (t_strides[sorted_index] !=
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/interface.cpp:145]: `and`?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/interface.cpp:145]: So this ended up being that a dimension != 1, but was broadcasted by setting stride = 0? I see how this check would have indicated it was contiguous even though it was really broadcasted.

**Review by csarofeen — COMMENTED** [review]
This makes sense to me, assigning Jie to confirm so not explicitly approving.

File: torch/csrc/jit/codegen/cuda/interface.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/interface.cpp:128]: I don't know if it's safe to omit the check when `sorted_index` is broadcasted by strides.
e.g. if `sorted_index` is considered an outer dimension with regarding to `inner_dim`, doesn't the broadcast scheme change that assumption?

File: torch/csrc/jit/codegen/cuda/interface.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/interface.cpp:145]: This is not safe any more is it?

If we mark a broadcasted dimension as contiguous, we might ended up collapsing the dimension to its inner dim, then we'll be indexing into broadcasted dimension and that's gonna be IMA.

**Comment by jjsjann123** [comment]
I'm feeling more comfortable with the alternative at #882 instead of this one.

Turns out that we run into two bugs, 1 with the chunk/concat trick that we copied from legacy fuser; 2. we have a slightly contiguity check as to what PE considers to be.

To be more specific on point 2. PE considers all dimension with stride == 1 to be contiguous. That's a little bit un-intuitive to me but I think it's safe to say so, (assume that we do not have overlapping memory.

**Comment by jjsjann123** [comment]
I'm closing this one since we decide to move forward with #882

**Pull request closed without merging**